src/modules/adaptors/Frax/DebtFTokenAdaptorV2.sol

```diff
@@ -0,0 +1,300 @@
+
+    /**
+     * @notice User deposits are NOT allowed into this position.
+     */
+    function deposit(uint256, bytes memory, bytes memory) public pure override {
+        revert BaseAdaptor__UserDepositsNotAllowed();
+    }
+
+    /**
+     * @notice User withdraws are NOT allowed from this position.
+     */
+    function withdraw(uint256, address, bytes memory, bytes memory) public pure override {
+        revert BaseAdaptor__UserWithdrawsNotAllowed();
+    }
+
+    /**
+     * @notice This position is a debt position, and user withdraws are not allowed so
+     *         this position must return 0 for withdrawableFrom.
+     */
+    function withdrawableFrom(bytes memory, bytes memory) public pure override returns (uint256) {
+        return 0;
+    }
+
+    /**
+     * @notice Returns the cellar's balance of the respective Fraxlend debtToken calculated from cellar borrow shares
+     * @param adaptorData encoded fraxlendPair (fToken) for this position
+     */
+    function balanceOf(bytes memory adaptorData) public view override returns (uint256) {
+        IFToken fraxlendPair = abi.decode(adaptorData, (IFToken));
+        return _toBorrowAmount(fraxlendPair, fraxlendPair.userBorrowShares(msg.sender), false, ACCOUNT_FOR_INTEREST);
+    }
+
+    /**
+     * @notice Returns `assetContract` from respective fraxlend pair, but this is most likely going to be FRAX.
+     */
+    function assetOf(bytes memory adaptorData) public view override returns (ERC20) {
+        IFToken fraxlendPair = abi.decode(adaptorData, (IFToken));
+        return ERC20(fraxlendPair.asset());
+    }
+
+    /**
+     * @notice This adaptor reports values in terms of debt.
+     */
+    function isDebt() public pure override returns (bool) {
+        return true;
+    }
+
+    //============================================ Strategist Functions ===========================================
+
+    // `borrowAsset`
+    /**
+     * @notice Allows strategists to borrow assets from Fraxlend.
+     * @notice `debtTokenToBorrow` must be the debtToken, NOT the underlying ERC20.
+     * @param fraxlendPair the Fraxlend Pair to borrow from.
+     * @param amountToBorrow the amount of `debtTokenToBorrow` to borrow on Fraxlend.
+     * NOTE: `borrowAsset` is the same btw v1 and v2 FraxlendPairs
+     */
+    function borrowFromFraxlend(IFToken fraxlendPair, uint256 amountToBorrow) public {
+        bytes32 positionHash = keccak256(abi.encode(identifier(), true, abi.encode(address(fraxlendPair))));
+        uint32 positionId = Cellar(address(this)).registry().getPositionHashToPositionId(positionHash);
+        if (!Cellar(address(this)).isPositionUsed(positionId))
+            revert DebtFTokenAdaptor__FraxlendPairPositionsMustBeTracked(address(fraxlendPair));
+
+        _borrowAsset(amountToBorrow, fraxlendPair);
+
+        // Check health factor is still satisfactory
+        uint256 _exchangeRate = _updateExchangeRate(fraxlendPair);
```

Comment:
Technically I think you could get away with just reading the exchange rate from the contract, since the exchange rate is updated when borrowing the asset. This is really only a gas efficiency though, and probs only saves you a little bit of logic gas (<100 gas)

Reply:
Did you want to remove this and simply read it? I have to triple-check that you're right on the reading the exchange rate only.